### PR TITLE
Generate default domain for Shoots

### LINKED ISSUE #544: Default shoot domain if provider != unmanaged
If the shoot's `.spec.dns.provider` field is not `unmanaged` then we should default the `.spec.dns.domain` field in case it is not provided by the user. Otherwise, the configuration does not make any sense and will lead to an error (#542).

Let's please default `.spec.dns.domain` with `<shootName>.<projectName>.<managed-domain>` where `<managed-domain>` is any of the registered default domains in the system.

The ShootDNSHostedZone admission plugin should take care of this.

### PR DESCRIPTION
**Pull request by timuthy**
**What this PR does / why we need it**:
In case the Shoot manifest does not have any domain configured in
`.spec.dns.domain` and `spec.dns.provider` field is not unmanaged,
the ShootDNSHostedZone admission plugin generates
a domain for it by applying the following pattern:
\<shootName\>.\<projectName\>.\<managed-domain\>

**Which issue(s) this PR fixes**:
Fixes #544 

**Release note**:

```improvement operator
Gardener does now generate a default domain name for newly created shoots that do not specify `.spec.dns.domain` but set `.spec.dns.provider!=unmanaged`. The domain will have the following scheme: `<shootName>.<projectName>.<managed-domain>`. This does only work if [default domains](https://github.com/gardener/gardener/blob/master/example/10-secret-default-domain.yaml) were registered in the system beforehand.
```

### CODE REVIEW

File: plugin/pkg/shoot/dnshostedzone/admission.go [hunk 1]
```diff
@@ -165,6 +176,52 @@ func (d *DNSHostedZone) Admit(a admission.Attributes) error {
...
+	}
+
+	if len(defaultDomain) > 0 && len(zoneID) > 0 && project != nil {
+		generatedDomain := fmt.Sprintf("%s.%s.%s", shoot.Name, project.Name, defaultDomain)
+		shoot.Spec.DNS.Domain = &generatedDomain
+		shoot.Spec.DNS.HostedZoneID = &zoneID
+	}
+}
+
+func getProject(shoot *garden.Shoot, projectLister gardenlisters.ProjectLister) (*garden.Project, error) {
```
Comment (rfranzke) [inline plugin/pkg/shoot/dnshostedzone/admission.go]: Could you please move this function to https://github.com/gardener/gardener/tree/master/plugin/pkg/utils? It is also used here: https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/validator/admission.go#L155-L164.
Please use the function on this place as well.

File: plugin/pkg/shoot/dnshostedzone/admission.go [hunk 2]
```diff
@@ -165,6 +176,52 @@ func (d *DNSHostedZone) Admit(a admission.Attributes) error {
...
+	projects, err := projectLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	for _, project := range projects {
+		if project.Spec.Namespace != nil && *project.Spec.Namespace == shoot.Namespace {
+			return project, nil
+		}
+	}
+	return nil, nil
```
Comment (rfranzke) [inline plugin/pkg/shoot/dnshostedzone/admission.go]:
```suggestion
	return nil, fmt.Errorf("Could not find project for namespace %q", shoot.Namespace)
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

**Comment by timuthy** [comment]
Thanks for your review @rfranzke! I've addressed your comments, PTAL.

**Review by rfranzke — APPROVED** [review]
/lgtm, nice!

**Pull request merged**